explorer/update.go

```diff
@@ -0,0 +1,189 @@
+package explorer
+
+import (
+	"sync"
+
+	"go.sia.tech/core/chain"
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+)
+
+// A Store is a database that stores information about elements, contracts,
+// and blocks.
+type Store interface {
+	AddSiacoinElement(sce types.SiacoinElement) error
+	AddSiafundElement(sce types.SiafundElement) error
+	AddFileContractElement(fce types.FileContractElement) error
+	RemoveElement(id types.ElementID) error
+	SiacoinElement(id types.ElementID) (types.SiacoinElement, error)
+	SiafundElement(id types.ElementID) (types.SiafundElement, error)
+	FileContractElement(id types.ElementID) (types.FileContractElement, error)
+	BlockFacts(height uint64) (BlockFacts, error)
+	AddBlockFacts(height uint64, facts BlockFacts) error
+	RemoveBlockFacts(height uint64) error
+	BlockHeight() (uint64, error)
+	SetBlockHeight(height uint64) error
```

Comment:
thinking about it some more, this API could cause performance headaches. If every `Add` or `Remove` call is fsync'ing the disk, and we call those functions dozens of times per update, then performance will be terrible.

I can think of two solutions:
- Add a `Commit` method, called when the `mayCommit` argument of `ProcessChainApplyUpdate` is set
- Consolidate the methods into just `AddChainUpdate` and `RevertChainUpdate`, moving much of the logic from `ProcessChainApplyUpdate`/`ProcessChainRevertUpdate` into the interface implementation.

The database we choose may end up influencing this decision, so let's postpone it for now.